### PR TITLE
[serve] Run `UserCallableWrapper` code on separate event loop (thread)

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Run all user code in the `UserCallableWrapper` on a separate event loop to avoid blocking the main replica loop.

After this PR is merged, I will be unblocked to productionize the [prototype](https://github.com/ray-project/ray/pull/42051) for late binding / single-RPC routing. This change is a requirement as we need to ensure the call to get the queue length/start executing a request (1) cannot hang due to user code blocking the event loop and (2) run on the same thread as each other.

This has a more general benefit in that it avoids having user code able to fully block our code in the replica (can help improve observability quite a lot) and makes the concurrency model more explicit. Currently some of the methods implicitly run on separate threads/event loops (using concurrency_group support) which makes it very easy to have subtle race conditions. For example user health check methods are not currently thread safe because they're actually called on a separate event loop.

I've removed the usage of `concurrency_group`s in this PR because they are no longer needed. This actually includes a minor bugfix for custom user `check_health` methods that can be considered a behavior change. Previously, the `check_health` method was run on a separate `asyncio` loop/thread as that method was in another concurrency group. Now, it runs on the same thread as the rest of the user code. This means users must avoid blocking the `asyncio` loop or else their custom health checks may time out. To avoid possibly causing problems for users without a custom health check, I've elected to skip interacting with the user code thread entirely in that case.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
@ray-project/ray-serve I'm planning to add unit tests for `UserCallableWrapper` in this PR before merging (because we can actually do that now 🎉), but PTAL at the general approach.

File: python/ray/serve/_private/replica.py [hunk 1]
```diff
@@ -665,22 +669,34 @@ async def perform_graceful_shutdown(self):
...
-    @ray.method(concurrency_group=REPLICA_CONTROL_PLANE_CONCURRENCY_GROUP)
     async def check_health(self):
-        await self._user_callable_wrapper.call_user_health_check()
+        # If the user provided a health check, call it via `user_callable_wrapper` on
+        # the user code thread. This means if user code blocks the event loop the health
+        # check may time out.
+        #
+        # To avoid this issue for basic cases without a user-defined health check, skip
+        # interacting with the `user_callable_wrapper` entirely.
+        if self._user_callable_wrapper.has_user_health_check:
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/replica.py:679]: Can we push this conditional into wrapper itself?

Comment (edoakes) [reply python/ray/serve/_private/replica.py:679]: good idea I think I'll need to do that for unit testing anyways

**Comment by edoakes** [comment]
# Microbenchmark results (run locally)

## HTTP noop latency

Master:
```
Latency (ms) for noop HTTP requests (num_replicas=1,num_requests=100):
count    100.000000
mean       3.569598
std        0.229940
min        3.265416
50%        3.505709
90%        3.846450
95%        4.068335
99%        4.227305
max        4.389583
```

This PR:
```
Latency (ms) for noop HTTP requests (num_replicas=1,num_requests=100):
count    100.000000
mean       3.626885
std        0.200631
min        3.311250
50%        3.569500
90%        3.900396
95%        3.993731
99%        4.145391
max        4.344958
dtype: float64
```

## Streaming handle throughput

Master:
```
DeploymentHandle streaming throughput (ASYNC) (num_replicas=1, tokens_per_request=1000, batch_size=10): 12632.2 +- 316.83 tokens/s
(ServeReplica:default:CallerDeployment pid=63933) Individual request quantiles:
(ServeReplica:default:CallerDeployment pid=63933)       P50=785.4314794999997
(ServeReplica:default:CallerDeployment pid=63933)       P75=835.9155417499999
(ServeReplica:default:CallerDeployment pid=63933)       P99=916.2086944999998
```

This PR:
```
DeploymentHandle streaming throughput (ASYNC) (num_replicas=1, tokens_per_request=1000, batch_size=10): 12119.74 +- 479.87 tokens/s
(ServeReplica:default:CallerDeployment pid=64031) Individual request quantiles:
(ServeReplica:default:CallerDeployment pid=64031)       P50=795.5596665
(ServeReplica:default:CallerDeployment pid=64031)       P75=907.079750249999
(ServeReplica:default:CallerDeployment pid=64031)       P99=946.3582387200001
```

## Streaming HTTP throughput

Master:
```
HTTP streaming throughput (num_replicas=1, tokens_per_request=1000, batch_size=10, use_intermediate_deployment=False): 247110.96 +- 2480.37 tokens/s
```

This PR:
```
HTTP streaming throughput (num_replicas=1, tokens_per_request=1000, batch_size=10, use_intermediate_deployment=False): 238696.0 +- 5205.73 tokens/s
```

File: python/ray/serve/_private/common.py [hunk 2]
```diff
@@ -657,15 +657,15 @@ class gRPCRequest:
...
 
 
 @dataclass
 class StreamingHTTPRequest:
     """Sent from the HTTP proxy to replicas on the streaming codepath."""
 
     pickled_asgi_scope: bytes
-    http_proxy_handle: ActorHandle
+    # Takes request_id, returns a pickled list of ASGI messages.
+    receive_asgi_messages: Callable[[str], Awaitable[bytes]]
```
Comment (edoakes) [inline python/ray/serve/_private/common.py:668]: This was refactored to take a generic callback for unit testing

**Comment by edoakes** [comment]
This is ready for review. Almost all of the LoC changed are added tests.

File: python/ray/serve/_private/replica.py [hunk 3]
```diff
@@ -1016,29 +1088,29 @@ async def call_user_method(
...
-            finally:
-                if hasattr(self._callable, "__del__"):
-                    del self._callable
+        if self._callable is None:
+            raise RuntimeError("initialize_callable must be called before destructor.")
+
+        # Only run the destructor once. This is safe because there is no `await` between
+        # checking the flag here and flipping it to `True` below.
+        if self._destructor_called:
+            return
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:1104]: This logic to only run the destructor once was not actually working properly (discovered via unit testing). I simplified with a boolean guard.

File: python/ray/serve/_private/http_util.py [hunk 4]
```diff
@@ -150,7 +150,8 @@ class MessageQueue(Send):
     """
 
     def __init__(self):
-        self._message_queue = asyncio.Queue()
+        # NOTE(edoakes): `deque` appends and pops are thread safe in CPython.
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/http_util.py:153]: Are there multiple threads using a single `MessageQueue`?

File: python/ray/serve/_private/http_util.py [hunk 5]
```diff
@@ -183,8 +187,8 @@ def get_messages_nowait(self) -> List[Message]:
         least one new message is available.
         """
         messages = []
-        while not self._message_queue.empty():
-            messages.append(self._message_queue.get_nowait())
+        while len(self._message_queue) > 0:
+            messages.append(self._message_queue.popleft())
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/http_util.py:191]: Is a `deque`'s `clear` method and a list's `extend` method threadsafe? If so, we could implement this as:

```suggestion
        messages.extend(self._message_queue)
        self._message_queue.clear()
```

**Review by shrekris-anyscale — COMMENTED** [review]
I haven't reviewed the tests yet, but overall this implementation looks good.

File: python/ray/serve/_private/http_util.py (hunk 4 above)
Comment (edoakes) [reply python/ray/serve/_private/http_util.py]: There are not. Sorry this was left over from a previous iteration of prototyping. Thread safety is handled using `run_coroutine_threadsafe` so only one asyncio loop/thread is accessing it (see `replica.py`).

File: python/ray/serve/_private/http_util.py (hunk 5 above)
Comment (edoakes) [reply python/ray/serve/_private/http_util.py:193]: I think these are equivalent?

Comment (shrekris-anyscale) [reply python/ray/serve/_private/http_util.py:193]: I think you're right ([source](https://www.scaler.com/topics/extend-vs-append-python/)). For some reason, I though `extend` was faster than adding elements one-by-one. We can leave it as-is.

File: python/ray/serve/_private/replica.py [hunk 6]
```diff
@@ -693,12 +703,45 @@ def __init__(
...
+        self._destructor_called = False
 
         # Will be populated in `initialize_callable`.
         self._callable = None
 
+        # Start the `_user_code_event_loop_thread` singleton if needed.
+        if self._user_code_event_loop_thread is None:
+
+            def _run_user_code_event_loop():
+                asyncio.set_event_loop(self._user_code_event_loop)
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/replica.py:715]: Would suggest to add a comment to elaborate why we need this one

File: python/ray/serve/_private/replica.py [hunk 7]
```diff
@@ -784,17 +831,36 @@ async def initialize_callable(self):
...
+        #
+        # To avoid this issue for basic cases without a user-defined health check, skip
+        # interacting with the user callable entirely.
+        if self._user_health_check is not None:
+            return await self._call_user_health_check()
+
+    @_run_on_user_code_event_loop
     async def call_reconfigure(self, user_config: Any):
+        if self._callable is None:
+            raise RuntimeError("initialize_callable must be called before reconfigure.")
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/replica.py:862]: Can extract to a method (like `_assert_initialized`)

File: python/ray/serve/_private/replica.py (hunk 3 above)
Comment (alexeykudinkin) [reply python/ray/serve/_private/replica.py:1104]: You don't need a separate boolean tho, you can just remove `self._callable` using it as idempotency token

**Review by alexeykudinkin — APPROVED** [review]
LGTM! Minor comments

File: python/ray/serve/tests/unit/test_http_util.py [hunk 8]
```diff
@@ -74,90 +65,79 @@ async def test_message_queue():
...
-        assert await asgi_receive_proxy() == {"type": "bar"}
+        loop = get_or_create_event_loop()
+        asgi_receive_proxy = ASGIReceiveProxy("", receive_asgi_messages)
+        receiver_task = loop.create_task(asgi_receive_proxy.fetch_until_disconnect())
 
-        ray.kill(actor)
-        with pytest.raises(ray.exceptions.RayActorError):
-            await asgi_receive_proxy()
+        try:
+            with pytest.raises(RuntimeError, match="maybe actor crashed?"):
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/unit/test_http_util.py:137]:
```suggestion
            with pytest.raises(RuntimeError, match="maybe actor crashed!"):
```

File: python/ray/serve/tests/unit/test_user_callable_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,637 @@
...
+            asyncio.ensure_future(user_callable_wrapper.call_destructor())
+            for _ in range(100)
+        ]
+    )
+    assert num_destructor_calls == 1
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("async_del", [False, True])
+async def test_no_user_health_check_not_blocked(async_del: bool):
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/unit/test_user_callable_wrapper.py:451]: `async_del` isn't used in this test.

File: python/ray/serve/tests/unit/test_user_callable_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,637 @@
...
+
+def basic_sync_generator(n: int, raise_exception: bool = False):
+    for i in range(n):
+        yield i
+
+        if raise_exception:
+            raise RuntimeError("uh-oh!")
+
+
+def basic_async_generator(n: int, raise_exception: bool = False):
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/unit/test_user_callable_wrapper.py:84]: Should this function be `async`?

**Review by shrekris-anyscale — APPROVED** [review]
Nice work, I left a few comments about the tests.

File: python/ray/serve/_private/replica.py (hunk 7 above)
Comment (edoakes) [reply python/ray/serve/_private/replica.py:862]: done

File: python/ray/serve/_private/replica.py (hunk 3 above)
Comment (edoakes) [reply python/ray/serve/_private/replica.py:1104]: I did this originally but that interferes with the check to assert that the callable has been initialized. Having a separate boolean is pretty clear anyways 🤷

File: python/ray/serve/_private/replica.py (hunk 6 above)
Comment (edoakes) [reply python/ray/serve/_private/replica.py:715]: good idea done

File: python/ray/serve/tests/unit/test_user_callable_wrapper.py (hunk 9 above)
Comment (edoakes) [reply python/ray/serve/tests/unit/test_user_callable_wrapper.py:451]: oops! surprised linter didn't yell

File: python/ray/serve/tests/unit/test_user_callable_wrapper.py (hunk 10 above)
Comment (edoakes) [reply python/ray/serve/tests/unit/test_user_callable_wrapper.py:84]: yes :)

**Pull request merged**